packages/graphqlsp/src/index.ts

```diff
@@ -237,15 +237,32 @@ function create(info: ts.server.PluginCreateInfo) {
 
             const span = { length: 1, start: node.end };
 
-            // TODO: if we have a typeImport but the name has been updated/is wrong
-            // we need to leave the "as x" typecast
-            const text =
-              source.text.substring(0, span.start) +
-              imp +
-              source.text.substring(
-                span.start + span.length,
-                source.text.length
-              );
+            let isTypeImportReplaced = false;
+            let text = '';
+            if (typeImport) {
+              const oldName = typeImport
+                .getText()
+                .split('.')
+                .pop()!
+                .replace(/(FragmentDoc|Document)$/, '');
+
+              const lengthDifference = oldName.length - name.length;
+              // Remove ` as ` from the beginning
+              imp = imp.slice(4);
+              text = source.text.replace(typeImport.getText(), imp);
+              span.start = span.start - lengthDifference;
+              span.length = imp.length + lengthDifference;
+              isTypeImportReplaced = true;
```

Comment:
Does span.start really need adjusting? Added a suggestion that might be a bit easier to read

```suggestion
  text =
    source.text.substring(0, span.start) +
    imp +
    source.text.substring(
      // We adjust the string here to account for the ` as ` (4) and the typeImport
      span.start + (span.length + typeImport.getText().length + 4),
      source.text.length
    );
    // We also have to adjust the span.length here to notify TS that we
    // are editing more than 1 character
    span.length = span.length + typeImport.getText().length + 4;
```

Reply:
Somehow this breaks it:

https://user-images.githubusercontent.com/571589/232542167-9f3a12d3-14c2-41ef-8f1a-ed2f70f2bd9e.mp4

Also the following empty line is swallowed

Reply:
I mean, then it's because of not adjusting the span.start

Reply:
So the reason for `lengthDifference` was that the length changes for the document and in the query itself.
Not sure if this is considered in your suggestion.

And you were right, in the code from the PR I could omit `span.start` without breaking things

Reply:
The whole code piece here is really confusing though, will try and read it again when I'm home

Reply:
Added a few comments to make it easier

```suggestion
              // We only want the oldExportName here to be present
              // that way we can diff its length vs the new one
              const oldExportName = typeImport
                .getText()
                .split('.')
                .pop()!

              // Remove ` as ` from the beginning,
              // this because getText() gives us everything
              // but ` as ` meaning we need to keep that part
              // around.
              imp = imp.slice(4);
              text = source.text.replace(typeImport.getText(), imp);
              span.length = imp.length + (oldExportName.length - exportName.length);
```

Reply:
Great, thanks for making it easier to comprehend